languages/js/src/dataFiltering.test.ts

```diff
@@ -0,0 +1,151 @@
+import { Oso } from './Oso';
+import { Relationship } from './dataFiltering';
+import 'reflect-metadata';
+import { Entity, PrimaryColumn, Column, createConnection } from 'typeorm';
+
+@Entity()
+export class Bar {
+  @PrimaryColumn()
+  id!: string;
+
+  @Column()
+  isCool!: boolean;
+
+  @Column()
+  isStillCool!: boolean;
+}
+
+@Entity()
+export class Foo {
+  @PrimaryColumn()
+  id!: string;
+
+  @Column()
+  barId!: string;
+
+  @Column()
+  isFooey!: boolean;
+}
+
+test('data filtering', async () => {
+  const connection = await createConnection({
+    type: 'sqlite',
+    database: `:memory:`,
+    entities: [Foo, Bar],
+    synchronize: true,
+    logging: false,
+  });
+  let bars = connection.getRepository(Bar);
+
+  let helloBar = new Bar();
+  helloBar.id = 'hello';
+  helloBar.isCool = true;
+  helloBar.isStillCool = true;
+
+  await bars.save(helloBar);
+
+  let foos = connection.getRepository(Foo);
+
+  let anotherFoo = new Foo();
+  anotherFoo.id = 'Another';
+  anotherFoo.barId = 'hello';
+  anotherFoo.isFooey = true;
+
+  await foos.save(anotherFoo);
+
+  const oso = new Oso();
+
+  function fromRepo(repo: any, name: string, constraints: any) {
+    let query = repo.createQueryBuilder(name);
+    let addedWhere = false;
+    for (let i in constraints) {
+      let c = constraints[i];
+      let clause;
+      switch (c.kind) {
+        case 'Eq':
+          {
+            clause = `${name}.${c.field} = :${c.field}`;
+          }
+          break;
+        case 'In':
+          {
+            clause = `${name}.${c.field} IN (:...${c.field})`;
+          }
+          break;
+      }
+      let param: any = {};
+      param[c.field] = c.value;
+      if (!addedWhere) {
```

Comment:
Tiny note: I think you can use `andWhere` exclusively -- you don't necessarily need an original `where` call. 

I was curious about this during your demo and looked up how it works in the typeorm querybuilder, and they only add the string `AND` if it's not the first where clause: https://github.com/typeorm/typeorm/blob/51b2a63d91d4e6a935e83f47fb958f4867e82e9c/src/query-builder/QueryBuilder.ts#L685-L696

Reply:
good find, changed it